base/compiler/tfuncs.jl

```diff
@@ -886,21 +963,20 @@ end
 
 function array_builtin_common_nothrow(argtypes, first_idx_idx)
     length(argtypes) >= 4 || return false
-    (argtypes[0] ⊑ Bool && argtypes[1] ⊑ Array) || return false
+    (argtypes[1] ⊑ Bool && argtypes[2] ⊑ Array) || return false
     for i = first_idx_idx:length(argtypes)
         argtypes[i] ⊑ Int || return false
     end
     # If we have @inbounds (first argument is false), we're allowed to assume we don't throw
-    (isa(argtypes[0], Const) && !argtypes[0].val) && return true
+    (isa(argtypes[1], Const) && !argtypes[1].val) && return true
     # Else we can't really say anything here
     # TODO: In the future we may be able to track the shapes of arrays though
     # inference.
     return false
 end
 
 # Query whether the given builtin is guaranteed not to throw given the argtypes
-function builtin_nothrow(@nospecialize(f), argtypes::Array{Any,1})
-    (f === tuple || f === svec) && return true
+function _builtin_nothrow(@nospecialize(f), argtypes::Array{Any,1}, @nospecialize(rt))
     if f === arrayset
```

Comment:
Why is `arrayset` the only case here that checks the number of arguments?

Reply:
Why do you expect that? If you think the tfunc below does that during inference, you would be wrong (it instead might not be able to analyze the expression for a few different reasons - mostly apply / isva related)